### PR TITLE
[HH] Add DCP16 PCIe attention collectives for Kimi K3

### PR DESCRIPTION
**Pull request by voipmonitor**
## Summary

- extend SparkInfer PCIe DCP attention gather/reduce from world sizes 2/4/8 to
  2/4/8/16 for Kimi K3 TP16/DCP16
- add fused paired projection/top-k transport, deferred hierarchical
  consumption, BF16x2 transport, and persistent small-M MoE barrier reuse used
  by the lossless target-only decode path
- retain model-free A/B/A correctness and performance harnesses, including the
  lane-allreduce proof of concept, even where a microbenchmark win did not
  improve current end-to-end decode
- harden the DCP A2A benchmark: validate launch geometry and batch capacity,
  make cleanup reachable after setup failures, run correctness oracles before
  timing, include LSE payload bytes, and record provenance plus raw timing data

Paired vLLM PR: local-inference-lab/vllm#238.

Umbrella issue: local-inference-lab/vllm#239.

## Why

Kimi K3 TP16/DCP16 halves target attention history per rank relative to DCP8,
but the existing PCIe attention channel rejected world size 16. DCP16 is needed
to retain a physical 1M target cache while leaving enough VRAM for either BF16
DSpark or the higher-throughput target-only runtime.

This is additive to merged PR #116. PR #116 supplied K3 DCP graph safety, FP8
gather correctness, empty-shard handling, and native E8M0 loading. This PR adds
world-size-16 support and the later communication optimizations.

## Validation

- Ruff and formatting pass on all changed Python files
- 98 CPU/fake-communication DCP A2A, hierarchical, and W4A16 configuration
  tests passed
- all modified CUDA extensions (DCP A2A, hierarchical all-reduce, and lane
  POC) compile against the published CUDA 13.2 runtime; the new two-mode
  direct-small-M CUDA-graph test is collected and ready for an unoccupied GPU
- DCP8/DCP16 eager and CUDA-graph attention oracles passed during tuning
- exact target graph pair improved from 117.88 to 98.30 us/layer in the original
  DCP16 launch sweep
- full original Kimi K3 target-only TP16/DCP16: 47.118 tok/s CC1 and
  204.559 tok/s aggregate CC8, with 1,460,937 physical FP8 KV tokens
- full original Kimi K3 TP16/DCP8 remains functional with 1,054,602 physical KV
  tokens; current dynamic-clock CC1 median is 38.054 tok/s
- all 24 inline CodeRabbit threads are resolved and the final CodeRabbit check
  passes; review fixes are in `804461d` and `34bb490`

## Reproduction

- Runtime matrix:
  [Kimi K3 HH profiles](https://github.com/local-inference-lab/rtx6kpro/blob/docs/kimi-k3-runtime-matrix-20260805/models/kimi-k3/hh-runtime-matrix-2026-08-05.md)
- Docker branch:
  [build/kimi-k3-hh-runtime-20260805](https://github.com/local-inference-lab/blackwell-llm-docker/tree/build/kimi-k3-hh-runtime-20260805)
- Image:
  `voipmonitor/vllm:kimi-k3-hh-runtime-pr238-pr118-20260805@sha256:e029cab81df9ef35cf55bf3caed6e62acaeabe87ad72a62722d10b5e07d3e66d`
- SparkInfer head: `34bb490b28fd0742006a611c83c6b9883ed3d453`
- vLLM head: `b7e203d0bb6e1456b858277a388726f93f0d1ff6`

AI assistance was used to implement, test, and organize this change. Martin Vit
is the human submitter responsible for reviewing and defending it.

## Summary by CodeRabbit

- **New Features**
  - Added PCIe communication support for 16-way tensor-parallel deployments.
  - Added paired all-gather operations, including optimized Kimi-K3 top-k routing.
  - Added configurable hierarchical all-reduce modes, including double-buffered and deferred-consumption execution.
  - Added flexible active-split handling for dense MLA workloads.
  - Added optional host-side barrier reset control for W4A16 workloads.

- **Benchmarking**
  - Expanded distributed benchmarks with correctness checks, latency statistics, CUDA Graph coverage, and JSON/CSV reporting.

- **Tests**
  - Added coverage for 16-way execution, new communication paths, configuration validation, active splits, and barrier-reset behavior.

### CODE REVIEW

**Comment by voipmonitor** [comment]
Clean-image E2E validation with vLLM PR #238 completed.

- SparkInfer head: `94c0cf0a08e93a2cef1616d5451d241c86049d2f`
- vLLM head: `a2433b38e98800ebc5e90efc11495779325f27bd`
- image: `voipmonitor/vllm:kimi-k3-hh-dspark-dcp16-pr238-pr118-20260804@sha256:a13b98f7c420c32c776872896a30c53867aa9dbab1b895c03d5ff15984179722`
- dense MLA target/draft TP16 preflight passed; DCP16 A2A warmed successfully
- physical KV cache: 1,057,049 tokens
- 512-token decode: 3.306 s after warmup, within 0.7% of the preserved reference runtime

The running workers use only `/opt/kimi-k3-hh/sparkinfer`, not the old dirty worktree.

File: benchmarks/benchmark_kimi_k3_dense_mla_splits.py [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+    if not parsed or any(item <= 0 for item in parsed):
+        raise argparse.ArgumentTypeError("expected comma-separated positive integers")
+    return parsed
+
+
+def _parse_args() -> argparse.Namespace:
+    parser = argparse.ArgumentParser(description=__doc__)
+    parser.add_argument("--device", default="cuda:0")
+    parser.add_argument("--page-size", type=int, default=768)
+    parser.add_argument("--max-cache-tokens", type=int, default=131_072)
```
[the remark this answers is not in the document]
Comment (voipmonitor) [reply benchmarks/benchmark_kimi_k3_dense_mla_splits.py]: Removed benchmark_kimi_k3_dense_mla_splits.py from this PR to keep the DCP16 collective change scoped. The correctness-gated benchmark is preserved on a separate optimization branch.

File: benchmarks/benchmark_kimi_k3_dense_mla_splits.py [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+    page_width = (args.max_cache_tokens + args.page_size - 1) // args.page_size
+    cache = torch.full(
+        (page_width, args.page_size, HEAD_DIM),
+        0.25,
+        dtype=FP8,
+        device=device,
+    )
+    q = torch.full((1, HEADS, HEAD_DIM), 0.125, dtype=FP8, device=device)
+    q_scale = torch.ones(1, dtype=torch.float32, device=device)
+    kv_scale = torch.ones(1, dtype=torch.float32, device=device)
```
[the remark this answers is not in the document]
Comment (voipmonitor) [reply benchmarks/benchmark_kimi_k3_dense_mla_splits.py]: Removed this unrelated dense-MLA benchmark from the DCP16 collective PR. The production FP8-KV K3 path intentionally uses an FP8 absorbed query with q_scale; a separate harness retains that exact contract.

File: benchmarks/benchmark_kimi_k3_dense_mla_splits.py [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+    page_table = torch.arange(page_width, dtype=torch.int32, device=device).view(1, -1)
+    cache_seqlens = torch.ones(1, dtype=torch.int32, device=device)
+    cu_seqlens_q = torch.tensor([0, 1], dtype=torch.int32, device=device)
+    reference_outputs: dict[int, torch.Tensor] = {}
+    all_results: list[dict[str, float | int]] = []
```
[the remark this answers is not in the document]
Comment (voipmonitor) [reply benchmarks/benchmark_kimi_k3_dense_mla_splits.py]: Removed the unrelated dense-MLA benchmark from this PR. Its separate optimization-branch version gates timing on the independent dense_mla.reference oracle, including non-unit scales and page-boundary cases.

File: benchmarks/benchmark_kimi_k3_dense_mla_splits.py [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+                "cosine_vs_first_plan": cosine,
+                "max_abs_vs_first_plan": max_abs,
+            }
+            all_results.append(result)
+            print(json.dumps(result), flush=True)
+
+        del graph, binding, output, scratch, plan
+        torch.cuda.empty_cache()
+
+    print(json.dumps({"results": all_results}), flush=True)
```
[the remark this answers is not in the document]
Comment (voipmonitor) [reply benchmarks/benchmark_kimi_k3_dense_mla_splits.py]: Removed the unrelated dense-MLA benchmark from this PR. The separately preserved harness records command, commit/worktree, GPU identity, correctness, raw samples, and ratio direction.

File: benchmarks/benchmark_pcie_dcp_a2a.py [hunk 5]
```diff
@@ -66,6 +84,11 @@ def _worker(rank: int, world_size: int, port: int) -> None:
         world_size=world_size,
     )
     dtype = torch.bfloat16
+    query_dtype = (
+        torch.float8_e4m3fn
+        if os.getenv("SPARKINFER_PCIE_DCP_A2A_QUERY_DTYPE", "bf16") == "fp8"
+        else torch.bfloat16
+    )
```
[the remark this answers is not in the document]
Comment (voipmonitor) [reply benchmarks/benchmark_pcie_dcp_a2a.py]: Fixed in 4fdc3e2: normalize the environment value and reject anything other than bf16 or fp8. The invalid-value path was exercised and raises the expected ValueError.

File: benchmarks/benchmark_pcie_hierarchical_tp12.py [hunk 6]
```diff
@@ -204,7 +205,10 @@ def _worker(rank: int, world_size: int, port: int) -> None:
 
 
 def main() -> None:
-    world_size = 12
+    parser = argparse.ArgumentParser(description=__doc__)
+    parser.add_argument("--world-size", type=int, choices=(12, 16), default=16)
+    args = parser.parse_args()
+    world_size = args.world_size
```
[the remark this answers is not in the document]
Comment (voipmonitor) [reply benchmarks/benchmark_pcie_hierarchical_tp12.py:234]: Fixed in 4fdc3e2: every CSV row now includes world_size, and the device-count error derives both TP label and required count from the selected world size. A TP16 GPU run passed.

File: benchmarks/reproduce_kimi_k3_tp16_direct_lifecycle.py [hunk 7]
```diff
@@ -55,6 +55,8 @@ def _parse_args() -> argparse.Namespace:
     parser.add_argument("--scan-start", type=int, default=0)
     parser.add_argument("--scan-stop", type=int)
     parser.add_argument("--scan-step", type=int, default=1)
+    parser.add_argument("--benchmark-warmups", type=int, default=0)
+    parser.add_argument("--benchmark-iterations", type=int, default=0)
```
[the remark this answers is not in the document]
Comment (voipmonitor) [reply benchmarks/reproduce_kimi_k3_tp16_direct_lifecycle.py]: Removed the newly added timing arguments from this PR and restored the existing lifecycle reproducer to its base behavior. The enhanced validated timing harness is preserved on a separate optimization branch.

File: benchmarks/reproduce_kimi_k3_tp16_direct_lifecycle.py [hunk 8]
```diff
@@ -288,6 +290,24 @@ def main() -> None:
...
+            graph.replay()
+        end.record()
+        end.synchronize()
+        elapsed_ms = start.elapsed_time(end)
+        print(
+            f"M=1 graph benchmark: iterations={args.benchmark_iterations}; "
+            f"total_ms={elapsed_ms:.6f}; "
+            f"per_layer_ms={elapsed_ms / args.benchmark_iterations:.6f}"
+        )
+
```
[the remark this answers is not in the document]
Comment (voipmonitor) [reply benchmarks/reproduce_kimi_k3_tp16_direct_lifecycle.py]: Removed the newly added timing path from this PR, so it can no longer report timings without the requested gate/provenance. The separate optimization harness performs graph-vs-eager correctness before timing and emits raw structured evidence.
